safelife/safelife_logger.py

```diff
@@ -0,0 +1,415 @@
+
+    def log_scalars(self, data, global_step=None, tag=None):
+        """
+        Log scalar values to tensorboard.
+
+        Parameters
+        ----------
+        data : dict
+            Dictionary of key/value pairs to log to tensorboard.
+        tag : str or None
+
+        """
+        self.init_logdir()  # init if needed
+
+        if not self.summary_writer:
+            return
+        tag = "" if tag is None else tag + '/'
+        if global_step is None:
+            global_step = self.cumulative_stats['training_steps']
+        for key, val in data.items():
+            if np.isreal(val) and np.isscalar(val):
+                self.summary_writer.add_scalar(tag + key, val, global_step)
+        self.summary_writer.flush()
+
+
+class RemoteSafeLifeLogger(BaseLogger):
+    """
+    Maintains a local interface to a remote logging object using ray.
+
+    The remote logging object is a ray Actor that does lightweight wrapping
+    of a SafeLifeLogger instance. This means that the same RemoteSafeLifeLogger
+    can be copied to different processes while maintaining a link to the same
+    actor, retrieving the same global state, and writing to the same open files.
+
+    Note that the ``cumulative_stats`` in the local copy will generally lag
+    what is available on the remote copy. It is only updated whenever an
+    episode is logged, and even then it is updated asynchronously.
+
+    Parameters
+    ----------
+    logdir : str
+        The directory in which to log everything.
+    config_dict : dict
+        A dictionary of options to pass to ``logging.config.dictConfig``
+        in the standard python logging library. Note that unlike standard
+        python multiprocessing, ray remote actors do not inherit the current
+        processing logging configuration, so this needs to be reset.
+    """
+    max_backlog = 50
+    update_interval = 0.01
+
+    @ray_remote
+    class SafeLifeLoggingActor(object):
+        def __init__(self, logger, config_dict):
+            self.logger = logger
+            logger.init_logdir()
+            if config_dict is not None:
+                logging.config.dictConfig(config_dict)
+
+        def log_episode(self, game, info, history, training):
+            self.logger.log_episode(game, info, history, training)
+            return self.logger.cumulative_stats
+
+        def log_scalars(self, data, step, tag):
+            self.logger.log_scalars(data, step, tag)
+
+        def update_stats(self, cstats):
+            self.logger.cumulative_stats = cstats
+
+    def __init__(self, logdir, config_dict=None, **kwargs):
+        if ray is None:
+            raise ImportError("No module named 'ray'.")
+        logger = SafeLifeLogger(logdir, **kwargs)
+        self.logdir = logdir
+        self.actor = self.SafeLifeLoggingActor.remote(logger, config_dict)
+        self._cstats = logger.cumulative_stats.copy()
+        self._promises = []
```

Comment:
Document this variable, perhaps?

Reply:
Eg: `_promises` stores references to remote updates to `cumulative_stats` that will be received in response to having sent a log item. There is no point exposing this state because there is in general no way to get up-to-date statistics to any thread, and therefore no benefit from knowing whether you're waiting for an update.